cou/steps/__init__.py

```diff
@@ -236,3 +244,82 @@ async def run(self) -> Any:
             return await self._task  # wait until task is completed
         except asyncio.CancelledError:  # ignoring asyncio.CancelledError
             logger.warning("Task %s was stopped unsafely.", repr(self))
+
+
+class UpgradePlan(BaseStep):
+    """Represents the upgrade plan.
+
+    This class is intended to be used as a higher-level group for actual upgrade steps, therefore
+    doesn't accept coroutine or parallel as inputs.
+    """
+
+    prompt: bool = False
+
+    def __init__(
+        self,
+        description: str = "",
+    ):
+        """Initialize upgrade plan.
+
+        :param description: Description of the step.
+        :type description: str
+        """
+        super().__init__(description=description, parallel=False, coro=None)
+
+    async def run(self) -> None:
+        """Run UpgradePlan.
+
+        UpgradePlan should not have contain any coroutine, so simply print a debug
+        message to demonstrate the noop.
+        """
+        logger.debug("No coroutine to run for %s", repr(self))
+
+
+class ApplicationUpgradePlan(UpgradePlan):
+    """Represents the plan for application-level upgrade.
+
+    This class is intended to be used as a group for application-level upgrade steps, therefore
+    doesn't accept coroutine or parallel as inputs.
+    """
+
+    prompt: bool = True
+
+
+class UpgradeStep(BaseStep):
+    """Represents the upgrade step."""
+
+    sub_step_prefix = "[upgrade] "
+    prompt: bool = False
+
+    @property
+    def description(self) -> str:
+        """Get the description of the UpgradeStep.
+
+        :return: description
+        :rtype: str
+        """
+        return self._description
+
+    @description.setter
+    def description(self, description: str) -> None:
+        """Set the description of the UpgradeStep.
+
+        :param description: description
+        :type description: str
+        :raises ValueError: When a coroutine is passed without description.
+        """
+        if not description and self._coro:
+            raise ValueError("Every coroutine should have a description")
+        self._description = self.sub_step_prefix + description
```

Comment:
Can we use `description_prefix` (or what will be the name of class variable) in `BaseStep`? It means that no overwritten for setter is needed here.